src/plan_ordered_append.c

```diff
@@ -0,0 +1,193 @@
+/*
+ * This file and its contents are licensed under the Apache License 2.0.
+ * Please see the included NOTICE for copyright information and
+ * LICENSE-APACHE for a copy of the license.
+ */
+
+#include <postgres.h>
+#include <nodes/relation.h>
+#include <nodes/value.h>
+#include <optimizer/paths.h>
+#include <optimizer/pathnode.h>
+#include <optimizer/tlist.h>
+#include <utils/builtins.h>
+#include <utils/typcache.h>
+
+#include "compat.h"
+#include "guc.h"
+#include "hypertable.h"
+#include "hypertable_restrict_info.h"
+#include "constraint_aware_append.h"
+#include "plan_ordered_append.h"
+
+/*
+ * Check if conditions for doing ordered append optimization are fulfilled
+ */
+bool
+ts_ordered_append_do_optimization(PlannerInfo *root, RelOptInfo *rel, Hypertable *ht, bool *reverse)
+{
+	/* check if optimizations are enabled */
+	if (ts_guc_disable_optimizations || !ts_guc_ordered_append)
+		return false;
+
+	/*
+	 * only do this optimization for hypertables with 1 dimension and queries
+	 * with an ORDER BY and LIMIT clause
+	 */
+	if (ht->space->num_dimensions != 1 || root->parse->sortClause == NIL || root->limit_tuples == -1.0)
+		return false;
+
+	/*
+	 * check that the first element of the ORDER BY clause actually matches
+	 * the first dimension of the hypertable
+	 */
+	{
+		SortGroupClause *sort = linitial(root->parse->sortClause);
+		TargetEntry *tle = get_sortgroupref_tle(sort->tleSortGroupRef, root->parse->targetList);
+		RangeTblEntry *rte = root->simple_rte_array[rel->relid];
+		TypeCacheEntry *tce;
+		char	   *column;
+
+		/* doublecheck rel actually refers to our hypertable */
+		Assert(ht->space->main_table_relid == rte->relid);
+
+		/*
+		 * check the ORDER BY column actually belonging to our hypertable
+		 * Unfortunately resorigtbl is not set for junk columns so we can't
+		 * doublecheck for those this way, but we are only dealing with single
+		 * tables at this level anyway so this should always match
+		 */
+		Assert(tle->resjunk == false || ht->space->main_table_relid != tle->resorigtbl);
+
+		/*
+		 * we only support direct column references for now
+		 */
+		if (!IsA(tle->expr, Var))
+			return false;
```

Comment:
just a note: you should look at sort_transform.c for transforming time-bucket etc.

Reply:
yes i already saw sort_transform_expr there which i would probably use here to support time_bucket